producer/writer/message.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"github.com/m3db/m3msg/generated/proto/msgpb"
+	"github.com/m3db/m3msg/producer"
+	"github.com/m3db/m3msg/protocol/proto"
+
+	"go.uber.org/atomic"
+)
+
+type message struct {
+	producer.RefCountedData
+
+	pb           msgpb.Message
+	meta         metadata
+	retryAtNanos int64
+	retried      int64
+	isAcked      *atomic.Bool
+}
+
+func newMessage() *message {
+	return &message{
+		retryAtNanos: 0,
+		retried:      0,
+		isAcked:      atomic.NewBool(false),
+	}
+}
+
+// Reset resets the message.
+func (m *message) Reset(meta metadata, data producer.RefCountedData) {
+	m.meta = meta
+	m.RefCountedData = data
+	m.retryAtNanos = 0
+	m.retried = 0
+	m.ToProto(&m.pb)
+	m.isAcked.Store(false)
+}
+
+// RetryAtNanos returns the timestamp for next retry in nano seconds.
+// This is NOT thread safe.
+func (m *message) RetryAtNanos() int64 {
```

Comment:
nit: Add WithLock suffix to this and the following thread unsafe methods since they are public?

Reply:
sounds good